### ISSUE TITLE
Alerts that refer to unready Providers don't recover automatically

### ISSUE THREAD

**Issue by seh** [fluxcd/notification-controller#204]
### Describe the bug

If you create a _Provider_ that doesn't become ready soon due to misconfiguration—such as its referenced _Secret_ not decrypting properly—and also create an _Alert_ that refers to that _Provider_, the _Alert_ does not notice when the _Provider_ later becomes ready.

### To Reproduce

1. Create a _Provider_ that references a _Secret_ that can't be decrypted (perhaps because the related _Kustomization_ decryption configuration is incorrect).
1. Note that the the _Provider_ is not ready.
1. Create an _Alert_ that refers to that _Provider_.
1. Note that the _Alert_ is not ready.
1. Correct the problem with the _Provider_ (perhaps allowing its referenced _Secret_ to be decrypted properly).
1. Note that the _Provider_ is now ready.
1. Watch the _Alert_ and note that it "never" becomes ready, or at least not within a reasonable amount of time.
1. Delete the _Alert_.
1. Create a replacement _Alert_.
1. Observe that the _Alert_ becomes ready almost immediately.

### Expected behavior

The _Alert_ should watch or poll at a low interval the referenced _Provider_, and its own status should change shortly after the _Provider_'s status changes.

### Additional context

- Kubernetes version: 1.19.9
- Git provider: GitHub
- Container registry provider: Docker Hub

Below please provide the output of the following commands:

<details>
<summary>flux tool output</summary>

```cli
% flux --version
flux version 0.14.0

% flux check
► checking prerequisites
✔ kubectl 1.21.1 >=1.18.0-0
✔ Kubernetes 1.19.9 >=1.16.0-0
► checking controllers
✔ helm-controller: deployment ready
► ghcr.io/fluxcd/helm-controller:v0.10.1
✔ kustomize-controller: deployment ready
► ghcr.io/fluxcd/kustomize-controller:v0.12.1
✔ notification-controller: deployment ready
► ghcr.io/fluxcd/notification-controller:v0.14.1
✔ source-controller: deployment ready
► ghcr.io/fluxcd/source-controller:v0.13.0
✔ all checks passed

% kubectl -n flux-system get all
NAME                                           READY   STATUS    RESTARTS   AGE
pod/helm-controller-85bfd4959d-9sgzz           1/1     Running   0          26m
pod/kustomize-controller-5687758989-nqr5d      1/1     Running   1          26m
pod/notification-controller-758d759586-zgkgs   1/1     Running   0          26m
pod/source-controller-85cd58f569-5xgnc         1/1     Running   0          26m

NAME                              TYPE        CLUSTER-IP       EXTERNAL-IP   PORT(S)   AGE
service/notification-controller   ClusterIP   100.65.113.231   <none>        80/TCP    4h3m
service/source-controller         ClusterIP   100.70.153.192   <none>        80/TCP    4h3m
service/webhook-receiver          ClusterIP   100.65.156.205   <none>        80/TCP    4h3m

NAME                                      READY   UP-TO-DATE   AVAILABLE   AGE
deployment.apps/helm-controller           1/1     1            1           4h3m
deployment.apps/kustomize-controller      1/1     1            1           4h3m
deployment.apps/notification-controller   1/1     1            1           4h3m
deployment.apps/source-controller         1/1     1            1           4h3m

NAME                                                 DESIRED   CURRENT   READY   AGE
replicaset.apps/helm-controller-85bfd4959d           1         1         1       4h3m
replicaset.apps/kustomize-controller-5687758989      1         1         1       4h3m
replicaset.apps/notification-controller-758d759586   1         1         1       4h3m
replicaset.apps/source-controller-85cd58f569         1         1         1       4h3m

[Logs elided.]
```
</details>

**Comment by SomtochiAma** [comment]
Hey @seh ,
I tried to recreate this bug but the alert and provider were both initialized(with Ready set to True) after the secret was decrypted successfully and the secret applied to the cluster.

**Comment by seh** [comment]
That's interesting. I haven't tried it again since first running into the problem, as my _Secrets_ are now built correctly, satisfying the _Providers_ from the start.

When you tried it, was the _Secret_ absent entirely, or was it present but with invalid content? In my case, I had forgotten to configure the Flux _Kustomization_ object's "spec.decryption" field properly, so I was winding up with the raw SOPS JSON content in my _Secret_.

**Comment by SomtochiAma** [comment]
In my case, it was absent entirely, I will try it with invalid content and see what happens.

**Comment by SomtochiAma** [comment]
Hey @seh, how were you able to apply the secret with the `sops` field in it(It gives a validation)? If I take it out, the kustomize-controller is unable to apply the secret(so it is still absent)
i.e 
```
apiVersion: v1
kind: Secret
metadata:
    name: slack-url
    namespace: flux-system
sops:
    kms: []
    gcp_kms: []
    azure_kv: []
    hc_vault: []
```

**Comment by seh** [comment]
> Hey @seh, how were you able to apply the secret with the `sops` field in it(It gives a validation error)?

I _thought_ I did, but perhaps I'm misremembering what had happened to put my _Provider_ to be stuck and not become ready. I know it was due to the missing decryption parameters, but I don't remember whether the _Secret_ object wound up existing or not.

If we can't reproduce the problem, we can close the issue. I mostly wanted to make sure the our controller the reconciles _Alert_ objects does watch _Providers_, and that the controller for _Providers_ watches _Secrets_. It behaved as if the the controller handling _Alerts_ wasn't noticing changes to the referenced _Provider_.
